### ISSUE TITLE
ReferenceEquality seems to break with switch expressions

### ISSUE THREAD

**Issue by jbrinegar** [google/error-prone#3038]
I managed to get this error when I used Intellij's "Replace with enhanced switch statement" inspection.

JDK: 17.0.2

```
      error-prone version: 2.11.0
      BugPattern: ReferenceEquality
      Stack Trace:
      com.google.common.util.concurrent.ExecutionError: java.lang.AssertionError: case visitor is implemented in SwitchBuilder
   	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2053)
   	at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
   	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3989)
   	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4950)
   	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4956)
   	at com.google.errorprone.dataflow.DataFlow.methodDataflow(DataFlow.java:176)
   	at com.google.errorprone.dataflow.DataFlow.expressionDataflow(DataFlow.java:236)
   	at com.google.errorprone.dataflow.nullnesspropagation.NullnessAnalysis.getNullness(NullnessAnalysis.java:59)
   	at com.google.errorprone.bugpatterns.AbstractReferenceEquality.getNullness(AbstractReferenceEquality.java:182)
   	at com.google.errorprone.bugpatterns.AbstractReferenceEquality.addFixes(AbstractReferenceEquality.java:103)
   	at com.google.errorprone.bugpatterns.AbstractReferenceEquality.matchBinary(AbstractReferenceEquality.java:78)
   	at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:449)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:512)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2166)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitAssignment(TreeScanner.java:676)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitAssignment(ErrorProneScanner.java:506)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitAssignment(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCAssign.accept(JCTree.java:2051)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:502)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitExpressionStatement(ErrorProneScanner.java:633)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitExpressionStatement(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1592)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:272)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:520)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1103)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitTry(TreeScanner.java:435)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitTry(ErrorProneScanner.java:845)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitTry(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCTry.accept(JCTree.java:1460)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:272)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:520)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1103)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitIf(TreeScanner.java:486)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitIf(ErrorProneScanner.java:654)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitIf(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1565)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:272)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:520)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1103)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:224)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:740)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:953)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:119)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:203)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:548)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:860)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:119)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:152)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:560)
   	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:150)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:614)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:60)
   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
   	at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
   	at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
   	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1394)
   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1341)
   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:933)
   	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
   	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
   	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
   	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
   	at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:89)
   	at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
   	at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
   	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:54)
   	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:39)
   	at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
   	at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
   	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
   	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
   	at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
   	at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
   	at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
   	at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
   	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:87)
   	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:56)
   	at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
   	at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
   	at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
   	at jdk.internal.reflect.GeneratedMethodAccessor14.invoke(Unknown Source)
   	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
   	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
   	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
   	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
   	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
   	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
   	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
   	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
   	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
   	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
   	at java.base/java.lang.Thread.run(Thread.java:833)
   Caused by: java.lang.AssertionError: case visitor is implemented in SwitchBuilder
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitCase(CFGTranslationPhaseOne.java:2200)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitCase(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCase.accept(JCTree.java:1357)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
   	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:119)
   	at jdk.compiler/com.sun.source.util.TreeScanner.visitSwitchExpression(TreeScanner.java:384)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCSwitchExpression.accept(JCTree.java:1390)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.translateAssignment(CFGTranslationPhaseOne.java:1518)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitAssignment(CFGTranslationPhaseOne.java:1510)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitAssignment(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCAssign.accept(JCTree.java:2051)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitExpressionStatement(CFGTranslationPhaseOne.java:2314)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitExpressionStatement(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1592)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:2062)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1103)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitTry(CFGTranslationPhaseOne.java:3086)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitTry(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCTry.accept(JCTree.java:1460)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:2062)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1103)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitIf(CFGTranslationPhaseOne.java:2741)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitIf(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1565)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:86)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:2062)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:198)
   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1103)
   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:60)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.process(CFGTranslationPhaseOne.java:428)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGTranslationPhaseOne.process(CFGTranslationPhaseOne.java:467)
   	at org.checkerframework.errorprone.dataflow.cfg.builder.CFGBuilder.build(CFGBuilder.java:70)
   	at com.google.errorprone.dataflow.DataFlow$2.load(DataFlow.java:127)
   	at com.google.errorprone.dataflow.DataFlow$2.load(DataFlow.java:95)
   	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3533)
   	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2282)
   	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2159)
   	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2049)
   	... 142 more
```

**Comment by carterkozak** [comment]
I believe this is resolved by specifying a higher version of `org.checkerframework:dataflow-errorprone` than `error_prone_core` requests by default. I've used 3.21.3 (current latest) with success.

**Comment by imoverclocked** [comment]
Also seems to be fixed with [errorprone 2.12.1](https://github.com/google/error-prone/releases/tag/v2.12.1) which pulls in `org.checkerframework:dataflow-errorprone:3.21.1` by default.

**Comment by jbrinegar** [comment]
confirmed fixed in 2.12.1
